File: ruby/2nd_ex/main.rb
def err(exc)
  raise StandardError.new(exc)
end

def init
  val = gets.to_i
  err("Invalid input") if val <= 0
  val
end

def cmp(arr, v)
  err("The volume of the watering can is not enough to water the flower") if v < arr.max
end

def counting_steps(arr, v)
  count = 0
  tmp = v

  arr.each_with_index do |value, i|
    if tmp < value
      tmp = v
      count += i * 2
    end
    count += 1
    tmp -= value
  end

  count
end

n = init

arr = []
n.times do
  arr << init
end

v = init

cmp(arr, v)

puts counting_steps(arr, v)

File: java/2nd_ex/main.java
import java.util.Scanner;

public class WateringPlants {
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        int n = init(scanner);
        
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = init(scanner);
        }
        
        int v = init(scanner);
        
        cmp(arr, v);
        
        System.out.println(countingSteps(arr, v));
    }
    
    private static int init(Scanner scanner) {
        int val = scanner.nextInt();
        if (val <= 0) {
            throw new IllegalArgumentException("Invalid input");
        }
        return val;
    }
    
    private static void cmp(int[] arr, int v) {
        for (int j : arr) {
            if (v < j) {
                throw new IllegalArgumentException("The volume of the watering can is not enough to water the flower");
            }
        }
    }
    
    private static int countingSteps(int[] arr, int v) {
        int count = 0;
        int tmp = v;
        
        for (int i = 0; i < arr.length; i++) {
            if (tmp < arr[i]) {
                tmp = v;
                count += i * 2;
            }
            count++;
            tmp -= arr[i];
        }
        
        return count;
    }
}
